Streamline migrating from HTTP to HTTPS

felixarntz opened the pull request:
**URL replacements from HTTP to HTTPS:**
* Introduces `wp_replace_insecure_home_url()` filter function that replaces the HTTP variant of `home_url()` with its HTTPS counterpart.
* Introduces `wp_should_replace_insecure_home_url()` (including `wp_should_replace_insecure_home_url` filter for customization) which determines whether the above function should run its replacement logic or not.
* Add `wp_update_https_migrated()` which hooks into `update_option_home` to set a `https_migrated` flag when `home_url()` is updated from HTTP to its HTTPS counterpart. If it is a "fresh site", there is no content to migrate, so the flag is set to `1`. Otherwise, it is set to `0` (i.e. URL replacement logic should kick in).
* Overall, URLs are only replaced if both `home_url()` and `site_url()` are using HTTPS and on the same domain. Special edge-cases (where e.g. the domain would differ between them) are not worth covering here since they would require additional logic, would make it more error-prone, and furthermore it is likely such sites are administered by fairly advanced WordPress users.

**Updating `home` and `siteurl` URLs to HTTPS:**
* Introduces `wp_update_urls_to_https()` which updates both options to their HTTPS counterparts. If as a result WordPress does not recognize the site as using HTTPS (e.g. due to the option being overridden by a constant), the update is reverted.
* Introduces `update_https` meta capability, which is given to users that have both `manage_options` and `update_core`.

**SIte Health integration:**
* Introduces `wp_get_update_https_url()` which returns the URL to learn more about updating HTTPS. It is customizable via either `WP_UPDATE_HTTPS_URL` environment variable or `wp_update_https_url` filter.
* Introduces `wp_get_direct_update_https_url()` which returns the "CTA" URL to actually update HTTPS, if set by the host. It can be set via either `WP_DIRECT_UPDATE_HTTPS_URL` environment variable or `wp_direct_update_https_url` filter.
* Integrates the above host URLs into the "HTTPS Status" Site Health section if they are specified.
* Adds a primary button if HTTPS is supported and can be enabled. The button goes either to the `wp_get_direct_update_https_url()` (if set) or triggers the URL update via the above `wp_update_urls_to_https()` function. Success/error feedback is provided accordingly upon redirect back to the Site Health page.

Trac ticket: https://core.trac.wordpress.org/ticket/51437

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

felixarntz commented:
Some screenshots of the different states of the "HTTPS Status" Site Health section with this PR:
**HTTPS not supported, no host support URL set:**
![Screen Shot 2021-01-13 at 5 05 54 PM](https://user-images.githubusercontent.com/3531426/104535204-98307780-55ca-11eb-8e71-5227a486f311.png)

**HTTPS not supported, host support URL set:**
![Screen Shot 2021-01-13 at 5 09 01 PM](https://user-images.githubusercontent.com/3531426/104535305-c6ae5280-55ca-11eb-87b9-8546524500dc.png)

**HTTPS supported, no host-specific direct update URL set:** (i.e. WordPress will update on its own)
![Screen Shot 2021-01-13 at 5 11 22 PM](https://user-images.githubusercontent.com/3531426/104535309-ca41d980-55ca-11eb-9cb8-46ecd6ff73d1.png)

**HTTPS supported, host-specific direct update URL set:**
![Screen Shot 2021-01-13 at 5 17 35 PM](https://user-images.githubusercontent.com/3531426/104535317-cdd56080-55ca-11eb-8bea-d891ceff6aee.png)

adamsilverstein commented:
@felixarntz This works well in my local testing, however I did find a couple of parts confusing from a user perspective.

* I tried setting my WordPress Address (wp-admin) to https, and my front end to http - a potentially common scenario, since https was at one point considered "expensive" and only worth running on the front end when context warranted it (eg. cart checkout):

![image](https://user-images.githubusercontent.com/2676022/105071964-7f004e80-5a42-11eb-96e7-6c04a44ae119.png)

When I visit site health, the warning is confusing:

![image](https://user-images.githubusercontent.com/2676022/105072045-9b03f000-5a42-11eb-86bd-937c35e30468.png)

In this case my "WordPress Address" **is using https**, but my "Site Address" is **not**. Maybe the text here can reflect that and help explain this to users. Something like:

_Your WordPress address is set to use HTTPS, but your Site Address is not set up to use HTTPS by default._

* My local site uses a self signed certificate, and I can see that is recorded successfully in the 'https_detection_errors' option, however -when I enable HTTPS for both WordPress and Site Address, the error detection is no longer used. So on Site Health, HTTPS shows under the passed tests. I feel like we should still want users when we detect that their certificate is invalid - what do you think?

One other question regarding the checks in `wp_update_https_detection_errors` - I'm not sure about caching the results in an option. Although the check takes some time to run, that seems fine for the Site Health page, and I'm worried we won't pick up changes when a user fixes the issue (for example by installing a valid certificate on their site). Do we really need to cache the results of these checks? The doc block says "This internal function is called by a regular Cron hook to ensure HTTPS support is detected and maintained.", but I didn't see that check?

felixarntz commented:
@adamsilverstein Thanks for the feedback.

> In this case my "WordPress Address" **is using https**, but my "Site Address" is **not**. Maybe the text here can reflect that and help explain this to users.

Good catch, I will update the messaging there to reflect that case.

> My local site uses a self signed certificate, and I can see that is recorded successfully in the 'https_detection_errors' option, however -when I enable HTTPS for both WordPress and Site Address, the error detection is no longer used. So on Site Health, HTTPS shows under the passed tests.

Great point, we already continue to have the check being made, so it's worth flagging such errors to the user in Site Health. I will update it so that, even if the site is already using HTTPS, this problem would be flagged. E.g. `Your site uses an invalid SSL certificate`.

> One other question regarding the checks in `wp_update_https_detection_errors` - I'm not sure about caching the results in an option. Although the check takes some time to run, that seems fine for the Site Health page

I agree it would probably be okay to not rely on a cached value for Site Health specifically, but this value is also used by a "regular" WordPress function that is available to call in any context, so I think relying on a cron hook makes more sense. The cron hook is added in `default-filters.php` as expected via `wp_schedule_https_detection` on `init`.

felixarntz commented:
@adamsilverstein Actually, on second though, all the feedback you left here applies more to https://core.trac.wordpress.org/ticket/47577 than to https://core.trac.wordpress.org/ticket/51437, the latter of which this is for. So I'll probably open a follow-up PR for https://core.trac.wordpress.org/ticket/47577 to make the changes based on your suggestions.

TimothyBJacobs commented:
If it isn't too much trouble, I think it'd probably be a good idea to do a fresh call in Site Health. That way a user can be sure the information they are seeing is up-to-date. It can be done as an async test, so it won't affect page load performance.

felixarntz commented:
@TimothyBJacobs Thanks! Fair point, we could call the update function there manually. That's something we can do as a follow-up commit for https://core.trac.wordpress.org/ticket/47577, as mentioned. Decoupled from that, it would be great to also get some feedback on this PR's approach, related to replacing URLs and streamlining the HTTPS _migration_.

TimothyBJacobs commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -1550,16 +1551,48 @@ public function get_test_https_status() {
+				if ( defined( 'WP_HOME' ) || defined( 'WP_SITEURL' ) ) {
+					$result['description'] .= sprintf(
+						'<p>%s</p>',
+						__( 'However, your WordPress Address is currently controlled by a PHP constant and therefore cannot be updated. You need to edit your <code>wp-config.php</code> and remove or update the definitions of <code>WP_HOME</code> and <code>WP_SITEURL</code>.' )
```
Should the wp-config and constant references be pulled out into a sprintf placeholders? Looking at `wp-admin/includes/network.php` for example.

TimothyBJacobs commented on src/wp-admin/site-health.php:
```diff
@@ -27,6 +29,23 @@
+	}
+
+	if ( ! wp_is_https_supported() ) {
+		wp_die( __( 'It looks like HTTPS is not supported for your website at this point.' ), 403 );
```
I think this `403` is meant for the previous `wp_die`?

TimothyBJacobs commented on src/wp-includes/https-migration.php:
```diff
@@ -0,0 +1,141 @@
+ * Updates the 'https_migrated' option if needed when the given URL has been updated from HTTP to HTTPS.
+ *
+ * If this is a fresh site, a migration will not be necessary, so the migration will be flagged as '1'. Otherwise, it
+ * will be flagged as '0'.
```
This kind of feels backwards to me. I think mainly because, AFAICT, this "migration" is never completed. It isn't like we set it to `0`, run a process to update all the URLS in the db, and then set it to `1`. It is only set to `1` if this is performed on a completely new site.

Based on the core usage, I would've expected something like `https_migration_required = 1`.

felixarntz commented:
@TimothyBJacobs Thanks for the feedback, I've addressed it. The previous feedback that was more related to https://core.trac.wordpress.org/ticket/47577 I've also addressed in a follow-up PR for that ticket, https://github.com/WordPress/wordpress-develop/pull/948.

TimothyBJacobs commented:
Code looks great to me, I should be able to give this a proper test on a site tomorrow.

Do URLs in CSS stylesheets cause mixed content warnings? Wondering if it might makes sense to also filter the customizer additional CSS rules.

felixarntz commented:
@TimothyBJacobs Good point, that's worth looking into. Will check tomorrow.

felixarntz commented:
@TimothyBJacobs I've refreshed this against latest `trunk` and also added the existing filter to `wp_get_custom_css`, so that HTTP versions of the `home_url()` are replaced with their HTTPS counterparts in custom CSS as well. cc @westonruter

TimothyBJacobs commented:
This looks good to me from the code. I unfortunately don't have an environment right now where I can test the actual migration to `https`.

felixarntz commented:
I've recorded a quick demo that shows the user flow of switching from HTTP to HTTPS, comparing how this is done on WordPress 5.6 vs WordPress 5.7: https://www.youtube.com/watch?v=wRVAY_Lwoco

felixarntz commented:
Committed via https://core.trac.wordpress.org/changeset/50131.
